### ISSUE TITLE
Jitsi loses password when unable to authenticate

### ISSUE THREAD

**Issue by cassioac** [jitsi/jitsi#150]
I use openfire server with AD authentication, and whenever I need the AD server becomes unavailable (due to a restart, for example), all connected Jitsi users are dropped due to lack of authentication (which is OK) but the cached saved password is lost, and users get a box like this:

https://cdn.pbrd.co/images/oqBiMt6.png

After the AD server is back, there are lame users (most of them) that doesn't remember to log in again and/or to check the "remember password" box, which results in lots of users becoming offline, needing technical support interaction for getting back online again.

Can't we have an option to make:

A: the checkbox continue to be checked and the password filled
B: an auto reconnect countdown
A/B Example: https://cdn.pbrd.co/images/or0t9Kv.png

C: a "stealth" mode, where jitsi keeps trying to reconnect as A/B example but the window popup asking for password is hidden. Maybe for this mode jitsi would change color on the task bar, meaning it's trying to connect.

Thanks

Cassio

**Comment by basildane** [comment]
This problem is really serious for us.
Another situation that causes it is when roaming users on laptops hibernate for the night.  When they wake up the laptops the next morning, AD is not immediately online, so Jitsi fails and forgets their password.

This happens EVERY SINGLE DAY.

50% of people on my server will not use Jitsi because of this.  Not out of spite, they just don't remember or want to reconfigure Jitsi every single morning.  Every time I need to IM someone, I need to call them on the phone to tell them they are "Offline".  This renders Jabber useless.  Everyone is always offline, so what good is it?

This is a critical problem.  I'm just glad I'm not the only one.

Jitsi needs to silently and automatically reconnect, and NOT forget it's configuration.

**Comment by fp2099** [comment]
My organization has the same problem.

**Comment by cassioac** [comment]
So, no update on this?

**Comment by basildane** [comment]
They're last message was to "use a different jabber client".  LOL.
So that's what I did.

**Comment by cassioac** [comment]
yeah I would rather also, but only jitsi has video, it's sad it has such poor support for other issues...

**Comment by cassioac** [comment]
I've been using pidgin for windows and adium for mac, where the video isn't necessary, both works way more stable...

**Comment by jitsi-developers** [comment]
Well, after searching a good alternative for skype i tried jitsi too. 
But it has a lot of issues and nobody helps.
Maybe the developers are just overloaded with tasks.
So use a different client how it is recommended :-D

Maybe jitsi will be better in some years, so try it later again.

Am 02.01.2016 um 19:36 schrieb cassioac:
>
> yeah I would rather also, but only jitsi has video, it's sad it has 
> such poor support for other issues...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jitsi/jitsi/issues/150#issuecomment-168415234>.
>
>
>
> _______________________________________________
> dev mailing list
> dev@jitsi.org
> Unsubscribe instructions and other list options:
> http://lists.jitsi.org/mailman/listinfo/dev



---
Diese E-Mail wurde von Avast Antivirus-Software auf Viren geprüft.
https://www.avast.com/antivirus

**Comment by damencho** [comment]
The described problem means that OF returns error with authentication failure, which means username and password is wrong, and there is no reason to try reconnecting. Normally, if just the connection is dropped reconnection will happen. 
Logs from such situation (the pcap which is normally in the logs archive) will help determine is there a way to distinct such cases and whether an easy resolution is available or OF sends the same error for authentication errors and the situation you describe.

**Comment by basildane** [comment]
damencho, not true in all cases.
In our case, when laptops wake up from sleep, the program tries to reconnect but the firewall proxy agent hasn't connected yet.  Jitsi gets a login failure NOT BECAUSE THE PASSWORD IS INVALID, but because the firewall has not yet authorized the session.  Then jitsi drops everything.
By the time the network is up and stable, jitsi has already thrown up a screen demanding you enter your username and password.  Users get annoyed by this after 50 times and they won't use it.

BTW, other programs do not do this.

**Comment by basildane** [comment]
Also, it's going to be difficult to get a pcap because the machine is not even fully up at this point yet.  I don't know how I would even start a trace while it is booting up.

**Comment by damencho** [comment]
Pcaps are automatically generated by jitsi. You can just send me the archive of logs from Options as described here: https://jitsi.org/logs. This will be helpful to determine the problem.

**Comment by basildane** [comment]
oh, in that case I'll try it.

**Comment by FreeSpirit9009** [comment]
I could provide pcap files from my machine; I am also affected by this bug. Would that help?

**Comment by mirohe** [comment]
I have the same problem.

When jitsi wants, to start the program (not always) asks me password.
It's a problem because I have 200 users using this program in my organization.

**Comment by cassioac** [comment]
I’ve migrated to pidgin, no more problems.

On 12 Dec 2016, at 08:41, mirohe <notifications@github.com<mailto:notifications@github.com>> wrote:


I have the same problem.

When jitsi wants, to start the program (not always) asks me password.
It's a problem because I have 200 users using this program in my organization.

**Comment by caiohennrique** [comment]
Has anyone managed to solve this? I'm also having the same problem

I believe it is related to network instability, when the desktop network becomes unstable it becomes more likely to lose the password. Is this a jitsi's native function? When the network is unstable it deletes the password from the file sip-communicator.properties

Does anyone have any idea where this function could be in the source code to deactivate it?

**Comment by basildane** [comment]
If anyone on the dev team is reading this, this is what you need to do to fix the problem correctly.

First, when Jitsi attempts to connect to the server to login, if it fails to connect, DO NOT ASSUME that means a wrong password.  The program should see if it can connect to server, not login, just connect and establish a TCP connection.

This may take several attempts, and up to 5 minutes.

After the connection is established THEN you try to login to the server.  Don't go logging in until you have a channel established.

The reason is that many users out there cannot just wake up and instantly connect to the Jabber host the second Jitsi comes up.  There are lots of factors that can delay the network connection.  Firewall proxies may not be established yet.  Also, at work we have NAC.  Network Access Control.  This prevents network communication until the NAC agent has communicated with a RADIUS server and authorized the computer to connect to the network.  This can take a minute.  Meanwhile, Jitsi starts up and immediately starts logging in, even though the network is down.  Doomed to fail.

These are not edge cases.  I have 65,000 users here at work that are all on NAC.  At home, our workstations cannot start talking until the firewall proxy agent is up.  Often, Jitsi starts up before the proxy agent, then it fails.

If anything I said is not clear, just say so and I will be happy to explain further.

**Comment by caiohennrique** [comment]
> if it fails to connect, DO NOT ASSUME that means a wrong password. The program should see if it can connect to server, not login, just connect and establish a TCP connection...
>If anything I said is not clear, just say so and I will be happy to explain further.

Yeah, I understand what you mean, that's what happens in my office.

 I do not understand how they do not solve this problem, I think it should not be difficult

What did you do to "solve" this problem? I thought about creating a gpo that would only read the file sip-communicator.properties

**Comment by basildane** [comment]
I solved it by switching my users who use hibernate to pidgin.
I took a look at the code and was going to try and fix it myself, but I couldn't get it to compile.

I really think this is just a one-line fix.  At least to stop it from clearing the credentials...

**Comment by caiohennrique** [comment]
> I really think this is just a one-line fix. At least to stop it from clearing the credentials

Really, should be something simple.
But I do not even know where to start looking for this, I do not know Java

**Comment by ibauersachs** [comment]
If it would be so easy, we would already have solved it. The problem is not that Jitsi assumes no connection=wrong credentials. @cassioac most likely uses or used Openfire. Jitsi only clears the credentials when the server reports that these are invalid, which also explains the situation with the AD server rebooting.
Simply not having a connection to the server definitively does not trigger clearing the password. I have never experienced this, but I would have on the countless WiFi hotspots which need an acknowledgment on a landing page. Or with the recent internet connection blackouts I had due to a broken cable modem.

If anyone wants to take a look at the "oneliner":  start debugging at https://github.com/jitsi/jitsi/blob/master/src/net/java/sip/communicator/impl/protocol/jabber/ProtocolProviderServiceJabberImpl.java#L644

If someone can me provide me with an environment that allows reproducing the issue I can and will look at it.

**Comment by cassioac** [comment]
all my jitsi -> openfire -> AD environments have this problem whenever AD goes offline and openfire can’t communicate with it.


On 12 Dec 2016, at 20:57, Ingo Bauersachs <notifications@github.com<mailto:notifications@github.com>> wrote:


If it would be so easy, we would already have solved it. The problem is not that Jitsi assumes no connection=wrong credentials. @cassioac<https://github.com/cassioac> most likely uses or used Openfire. Jitsi only clears the credentials when the server reports that these are invalid, which also explains the situation with the AD server rebooting.
Simply not having a connection to the server definitively does not trigger clearing the password. I have never experienced this, but I would have on the countless WiFi hotspots which need an acknowledgment on a landing page. Or with the recent internet connection blackouts I had due to a broken cable modem.

If anyone wants to take a look at the "oneliner": start debugging at https://github.com/jitsi/jitsi/blob/master/src/net/java/sip/communicator/impl/protocol/jabber/ProtocolProviderServiceJabberImpl.java#L644

If someone can me provide me with an environment that allows reproducing the issue I can and will look at it.

**Comment by ibauersachs** [comment]
@cassioac I'm not sure what Openfire is reporting in that case. It's possible that we do mishandle something, but a service unable to authenticate any user shouldn't accept logins at all. And this case is different from the others mentioned here with the NACs and late network startup.

**Comment by cassioac** [comment]
Please let me know if you want me to provide any kind of log

Sent from my iPhone

On 13 Dec 2016, at 19:22, Ingo Bauersachs <notifications@github.com<mailto:notifications@github.com>> wrote:


@cassioac<https://github.com/cassioac> I'm not sure what Openfire is reporting in that case. It's possible that we do mishandle something, but a service unable to authenticate any user shouldn't accept logins at all. And this case is different from the others mentioned here with the NACs and late network startup.

-

**Comment by ibauersachs** [comment]
@cassioac I'll have a look if you have a complete set of working vs. non-working logs. You can send them to me directly if you don't want them floating around publicly.

**Comment by basildane** [comment]
I haven't done a trace directly, but consider this possibility.
Laptop wakes up from sleep.
Firewall proxy begins to login.
Jitsi starts up and tries to login.  Firewall proxy HAS NOT finished setting up yet.
Jitsi gets a 401 something because it is unauthorized to make any network requests - because the firewall is forbidding everything until proxy is up.
Jitsi now deletes credentials and asks user to reenter them.
Proxy has now connected and network connectivity is running.
User scratches head and wonders what just happened.

**Comment by ibauersachs** [comment]
@basildane If your firewall proxy hijacks a TLS secured connection and injects a SASL not-authorized response: I'm impressed. And it wouldn't be Jitsi's problem because in this case it would behave as expected.
If there's no connection to the XMPP server at all, then nothing happens. Just like in the case with the WiFi hotspot I described earlier. I can't investigate anything without a proper set of logs from Jitsi.

**Comment by ibauersachs** [comment]
Any chance someone can sends me the .pcap files from an affected client? If not, I'm going to close this issue.

**Comment by LesterJitsi** [comment]
Well, I'm totally ignorant but created an account here as I'm unable to login to my two Jabber accounts (keep getting invalid password message). My problem isn't exactly the one this thread describes, but I think may be related. I'm mailing you my .pcap files. Hope this helps.

**Comment by ibauersachs** [comment]
@LesterJitsi http://lists.jitsi.org/pipermail/users/2016-April/011051.html

As I got no other logs, I'm closing this issue.

**Comment by MountainMaster** [comment]
Same problem for me.
Getting logged out from SIP Account every other day.

If you are willing to reopen i can provide Logs.

**Comment by mirohe** [comment]
I have also seen that if you install a new version of java (always having another installed in the system), or uninstall and install again, it always loses the authentication.

Please, I have 200 users with this problem

**Comment by ibauersachs** [comment]
@MountainMaster if you can't attach the logs while the issue is closed, please send them directly to me.

@mirohe Your issue is different. Switching Java versions changes the supported ciphers (AES 128 vs AES 256). Jitsi uses AES 256 to encrypt the passwords in the settings if available. If it's not available (which is the case by default in Oracle's Java), it'll use AES 128. Other than generally switching to AES 128, there's nothing we can do.

**Comment by mirohe** [comment]
Yes I understand, but also when jitsi wants, to start the program (not always) prompts me for password, and it is not a problem installing or changing the java version.
I wrote this above.

**Comment by jhertel** [comment]
The problem is still there: When Jitsi is unable to contact the server (or maybe even gets rejected by the server), it seems to incorrectly conclude that the password must be wrong (it never is), and perhaps even deletes the remembered password (really?), as it keeps asking even when the network connection comes back and you restart Jitsi after that.

And even _if_ the case is that the server erroneously rejects the password, Jitsi should not conclude that the password is _actually_ wrong – it should still suggest the previously used password in the dialog box (dotted out of course in a proper password field, maybe even using a fake temporary password of the same length to avoid password peekers), so the user doesn't have to type in the same password again and again.

**Comment by jhertel** [comment]
I don't feel like supplying pcap log files, as they contain sensitive data and shouldn't _really_ be related to this. The problem is about how Jitsi behaves generally in certain circumstances, such as seemingly forgetting the password if it gets rejected by the server. In my mind, Jitsi should _never_ forget the password unless the user specifically asks Jitsi to forget. So even if the server actively rejects and says unauthorized, Jitsi should _not_ forget the password, as the server might be wrong (there are bugs in server software too). If there is disagreement about this behaviour (forgetting the password or not), it should be an option.

Even though I will not supply the entire pcap files, you are most welcome to ask me questions about the pcap log file for a session where this has actually happened. Then I'll look up that specific information. 🙂 I saved one of those pcap files in a safe place.

**Comment by ibauersachs** [comment]
@jhertel The .pcap's obviously don't contain the actual password, but I understand that you don't want to share other details.

Jitsi is erasing the password because it is currently the only way to bring up the password dialog during the next logon attempt. I don't disagree with you that this feels wrong, but I also don't have an easy fix/workaround. There's at least these cases to consider:
- The password is actually invalid
- The password is valid, but has expired and cannot be used to login anymore
- The password is valid, but the server says otherwise
Simply retrying (as has been suggested either here or in another issue, I can't remember) might trigger an account lockout due to repeated retries.

Since the problem never occured to myself, I was unable to capture a proper trace to at least soften the symptoms. However, if a server is known to behave faulty (e.g. replying `not-authorized` instead of `temporary-auth-failure`, which is what _might_ happen in some cases, e.g. when an LDAP server/AD domain controller is unavailable), then that should be fixed too.

Those who can compile Java and have the problem repeatedly: does it also happen with the smack4-2 branch? The Smack update might bring some relief. Otherwise, help on implementing a change would be very welcome!
